lib/engine/game/base.rb

```diff
@@ -597,16 +598,46 @@ def initialize_actions(actions)
         end
         @redo_possible = active_undos.any?
         @loading = false
+        @loading_internal = false
       end
 
       def process_action(action)
         action = action_from_h(action) if action.is_a?(Hash)
+
         action.id = current_action_id + 1
         @raw_actions << action.to_h
         return clone(@raw_actions) if action.is_a?(Action::Undo) || action.is_a?(Action::Redo)
 
         @actions << action
 
+        # Process the action we came here to do first
+        process_single_action(action)
+
+        # Process any derived actions
+        # derived actions cannot be undos so we don't need to worry about state
+        action.derived.each { |a| process_single_action(a) }
+        any_derived = !action.derived.empty?
+
+        unless @loading_internal
```

Comment:
is this whole section necessary? 

can add_action(x) just mutate the state of the actions?

and then we just do 

while (action = action.derived_actions.pop)

Reply:
we can rename to auto action

Reply:
not really, if you have chained auto actions this wouldn't work since the step can change

Reply:
i think it should work if it's done in a while looop?

Reply:
as mentioned in chat pop mutates the list which is needed outside to actually store the auto_actions.

I'll try and see if there's any tidying up but my initial attempts don't seem to be going anywhere